my-megaparsec-project/src/Statement.hs

```diff
@@ -0,0 +1,155 @@
+  | Read Text
+  deriving (Show, Eq)
+
+evalStmt :: Stmt -> State (Subst, [Int]) [Int]
+evalStmt (Assign var e) = do
+  (subst, input) <- get
+  let e' = evalExpr e subst
+  put (M.insert var e' subst, input)
+  return []
+
+evalStmt (Seq s1 s2) = do
+  out1 <- evalStmt s1
+  out2 <- evalStmt s2
+  return (out1 ++ out2)
+
+evalStmt (Write e) = do
+  (subst, input) <- get
+  return [evalExpr e subst]
+
+evalStmt (Read var) = do
+  (subst, input) <- get
+  case input of
+    []     -> error "No more input values!"
+    (x:xs) -> do
+      put (M.insert var x subst, xs)
+      return []
+
+evalStmt (While cond body) = do
+  (subst, input) <- get
+  let condVal = evalExpr cond subst
+  if condVal < 0
+    then throw (IllegalArgumentException "While condition evaluated to negative value")
+    else if condVal == 0
+      then return []
+      else do
+        put (subst, input)
+        out1 <- evalStmt body
+        out2 <- evalStmt (While cond body)
+        return (out1 ++ out2)
+
+prettyPrintStmt :: Stmt -> String
+prettyPrintStmt (Assign var e) = show var ++ " = " ++ prettyPrintExpr e ++ ";"
+prettyPrintStmt (Seq s1 s2) = prettyPrintStmt s1 ++ "\n" ++ prettyPrintStmt s2
+prettyPrintStmt (While cond body) = "while (" ++ prettyPrintExpr cond ++ ") {\n" ++ prettyPrintStmt body ++ "\n};"
+prettyPrintStmt (Write expr) = "print(" ++ prettyPrintExpr expr ++ ");"
+prettyPrintStmt (Read var) = show var ++ " = readInt();"
+
+{-
+My aim is to start parsing simple stgatement C like language:
+Simple Fibonacci Program:
+
+n = readInt();
+a = 0;
+b = 1;
+i = 1;
+while ((n - i)) {
+  next = (a + b);
+  a = b;
+  b = next;
+  i = (i + 1);
+};
+print(b);
+
+actually readInt and print not "reads and writes Ints",
+ but takes and put thems into In / Out list"
+-}
+
+parseBody :: Parser Stmt
+parseBody = curly parseProgram
+
+readInt :: Parser Text
+readInt = lexeme $ do
+  _ <- keyword "readInt"
+  _ <- symbol '('
+  _ <- symbol ')'
+  return "readInt()"
+
+
+parseAssign :: Parser Stmt
+parseAssign = do
+  var <- identifier
+  _ <- symbol '='
+  Assign var <$> parseExpr
+
+parseWhile :: Parser Stmt
+parseWhile = do
+  _ <- keyword "while"
+  cond <- parens parseExpr
+  While cond <$> parseBody
+
+parseSeq :: Parser Stmt
+parseSeq = do
+  first <- parseStmt
+  Seq first <$> parseProgram
+
+parseWrite :: Parser Stmt
+parseWrite = do
+  _ <- keyword "print"
+  expr <- parens parseExpr
+  return $ Write expr
+
+parseRead :: Parser Stmt
+parseRead = do
+  var <- identifier
+  _ <- symbol '='
+  _ <- readInt
+  return $ Read var
+
+-- | Parse a semicolon and optionally consume newlines after it
+semicolon :: Parser ()
+semicolon = do
+  _ <- symbol ';'
+  _ <- many (symbol '\n')  -- Consume any number of newlines after semicolon
+  return ()
+
+
+parseStmt :: Parser Stmt
+parseStmt = do
+  stmt <- try parseRead <|> parseWrite <|> parseWhile <|> try parseAssign
+  _ <- semicolon
+  return stmt
+
+parseProgram :: Parser Stmt
+parseProgram = try parseSeq <|> parseStmt
```

Comment:
If your program is 1 statement long, you will re-parse it. I'd just treat a sequence of statements as a list